Add bio field to account

cahrens opened the pull request:
@andy-armstrong I think this is ready for your review. I have a sandbox with the user API enabled so you can play around with updating the bio field.

http://cahrens.m.sandbox.edx.org/api/user/v0/accounts/verified

This replaces #7331.

andy-armstrong commented on common/djangoapps/student/migrations/0046_auto__add_entranceexamconfiguration__add_unique_entranceexamconfigurat.py:
```diff
@@ -121,9 +121,11 @@ def backwards(self, orm):
             'change_date': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'blank': 'True'}),
             'changed_by': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']", 'null': 'True', 'on_delete': 'models.PROTECT'}),
-            'dashboard_tracking_code': ('django.db.models.fields.TextField', [], {'blank': 'True'}),
+            'company_identifier': ('django.db.models.fields.TextField', [], {}),
```
Why is this migration being updated by your change? That seems odd...

cahrens replied:
It's not-- this is the Cale commit that is necessary for migrations to work. The comment got lost in the PR change.

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -68,6 +68,7 @@ def create_mock_profile(self, user):
         legacy_profile.year_of_birth = 1900
         legacy_profile.goals = "world peace"
         legacy_profile.mailing_address = "Park Ave"
+        legacy_profile.bio = "Tired mother of twins"
```
hahahah!

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -68,6 +68,8 @@ class AccountView(APIView):
 
                 * goals: null or textual representation of goals
 
+                * bio: null or textural representation of user biographical information ("about me")
```
Should there be any restrictions on what can go in the bio? I assume that any Unicode character is valid. Should we say that HTML is not supported (as in it will be escaped into the bio to avoid script injection). Did we decide that there is essentially no size limit?

cahrens replied:
Good questions-- @explorerleslie what do you think? Should we do something to escape or expunge HTML?

I think there is essentially no size limit.

explorerleslie replied:
No HTML allowed in bio. It might be nice to enforce a size limit so people don't type manifestos, but not required.

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -93,7 +94,7 @@ def _verify_full_shareable_account_response(self, response):
         self.assertIsNone(data["time_zone"])
         self.assertIsNone(data["languages"])
-        self.assertIsNone(data["bio"])
+        self.assertEqual("Tired mother of twins", data["bio"])
```
Maybe there should be a test to verify clearing of the bio and making sure that the change sticks, i.e. that empty string updates are not treated as no-ops.

cahrens replied:
I'll look into that.

cahrens replied:
Actually, I already had test coverage for that at the end of test_patch_account:
````
# If there are no values that would fail validation, then empty string should be supported.
self.send_patch(self.client, {field: ""})

get_response = self.send_get(self.client)
self.assertEqual("", get_response.data[field])

andy-armstrong commented:
:+1: 

Note that it looks like you are suffering from a flaky test.

andy-armstrong commented:
I tried setting my bio to some Chinese text and it came back looking escaped. I don't know whether this is working correctly, but it would be good to have some unit tests around such characters.

This is the text I tried, btw:

壓是進界推日不復女；可給比少有而費有問頭幾深市如因這當度買我本選路們部的國，裡常濟；得強安增電世充歡最兩市那作得毛國：機境腦引筆反！黃去區立人機兒出聽必終化那？那人雲美需小華無點下大先懷，民關身酒長醫進目就內水，的古可後戰緊裝力後師手然他我裡，們明小去兒小運快年技元金學時，英長公，上治你務你資法見推舞院我？電先先傳鄉文？使們式銀。了新然因關關理上是滿時四！裡論次一！樂大想示！件地然氣不以斷鄉雙記什山石麗學！生創不活器麼己有供西麼代義之太有這量本合道共雜些出前社意不來講們己五輪也。達灣先理見，足集力建什來那的土任史自造：文們部入。是勢那家了獎商心政童來大己處人想力業麼去們配靈利樣著從金管團！因中的論富害語代說起自陽裡好高深後只，加場不跟到；爸明運現單學士人很下是讀保都候用我似水她樂他演得富有女好是情說養收環知，到聲這品一重地林好山是臺教地心是策心。構推負高信工理戰家報卻教排不驚花人

andy-armstrong commented:
FYI, I got it from here. I've no idea what it says!

http://www.richyli.com/tool/loremipsum/

cahrens commented:
Good point-- I will look into how character encoding is handled and add tests.

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -68,6 +68,12 @@ class AccountView(APIView):
 
+                * bio: null or textural representation of user biographical information ("about me")
+
+            For all text fields, clients rendering the values should take care to HTML escape them to avoid
```
@symbolist A couple of things to note and test in your account and profile pages--

1) Whatever text is supplied will be stored as-is. So you should make sure to HTML-escape the text on display (so if the user typed in HTML, what they see will be the literal HTML code, not what they would see if the HTML was rendered as HTML). This is necessary to avoid script injections.

2) We think that unicode characters will update and display properly. Please test some non-ASCII characters and let us know if there is an issue.

cahrens commented:
@andy-armstrong I added a couple of test cases and some explanatory text (and I squashed). I'm planning to merge this to the feature branch once tests pass.

Note that there is an extra commit that needs to be removed at some point, because there was a missing migration file. Cale has a PR with this change in it, so assuming his merges first, we will delete this commit.
